src/OpenAddingIterator.php

```diff
@@ -0,0 +1,83 @@
+<?php
+
+namespace MaxGoryunov\SavingIterator\Src;
+
+use ArrayAccess;
+use Iterator;
+
+/**
+ * Adding iterator which stores values in a user provided iterator.
+ * @template TKey
+ * @template TValue
+ * @implements AddingIterator<TKey, TValue>
+ */
+final class OpenAddingIterator implements AddingIterator
+{
+
+    /**
+     * Ctor.
+     * 
+     * @phpstan-param Iterator<TKey, TValue>&ArrayAccess<TKey, TValue> $added
+     * @param Iterator&ArrayAccess $added iterator with stored values.
+     */
+    public function __construct(
+        /**
+         * Iterator with stored values.
+         * 
+         * @phpstan-var Iterator<TKey, TValue>&ArrayAccess<TKey, TValue>
+         * @var Iterator&ArrayAccess
+         */
+        private Iterator|ArrayAccess $added
```

Comment:
Currently PHP does not support intersection types so it is not possible to write this otherwise.